### PR TITLE
Ability to set access rights per user (Issue-2879)

### LINKED ISSUE #2879: Per user access rights for Layman data
# Feature Request

**Is your feature request related to a problem? Please describe.**
We currently allow setting only private/public access rights for layers or compositions published to Layman. It is supported to set permissions for specific users.

**Describe the solution you'd like**
Allow to specify users with read/write permissions for layers and compositions.

**Additional context**
[Layman docs](https://github.com/LayerManager/layman/blob/master/doc/security.md#publication-access-rights)

### PR DESCRIPTION
**Pull request by DailisLangovskis**
## Description

This PR allows the choose between setting Layman access_rights for everyone or per user

## Related issues or pull requests

closes #2879 

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Dependency updates
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)
- [ ] I am unsure (we'll look into it together)

## Do you introduce a breaking change?

- [ ] Yes
- [x] No
- [ ] I am unsure (no worries, we'll find out)

## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the [MIT License]
- [x] I have followed the [guidelines for contributing](https://github.com/hslayers/hslayers-ng/blob/master/CONTRIBUTING.md)
- [x] The proposed change fits to the content of the [code of conduct](https://github.com/hslayers/hslayers-ng/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have added or updated tests and documentation, and the test suite passes (run `npm test` locally)
- [ ] I'm lost; why do I have to check so many boxes? Please help!

### CODE REVIEW

File: projects/hslayers/src/assets/locales/en.json [hunk 1]
```diff
@@ -785,7 +787,9 @@
     "syncErrorDialog": {
       "errorWhenSyncing": "Error when syncing"
     },
-    "writeAccessRights": "Write access rights"
+    "writeAccessRights": "Write access rights",
+    "preUser": "Per user",
```
Comment (jmacura) [inline projects/hslayers/src/assets/locales/en.json:791]: "perUser", please

File: projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts [hunk 2]
```diff
@@ -1,27 +1,170 @@
...
+import {HsLaymanService} from '../../../components/save-map/layman.service';
+import {HsLogService} from './../../log/log.service';
 import {accessRightsModel} from '../../../components/add-data/common/access-rights.model';
 
+type LaymanUser = {
+  username: string; // Username of the user.
+  screen_name: string; // Screen name of the user.
+  given_name: string; // Given name of the user.
+  family_name: string; // Family name of the user
+  middle_name: string; // Middle name of the user
```
Comment (jmacura) [inline projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts:17]: Could you change these property names to camelCase?

File: projects/hslayers/src/assets/locales/en.json (hunk 1 above)
Comment (DailisLangovskis) [reply projects/hslayers/src/assets/locales/en.json]: Sorry about that!

File: projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts (hunk 2 above)
Comment (DailisLangovskis) [reply projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts]: Done!

File: projects/hslayers/src/assets/locales/en.json (hunk 1 above)
Comment (jmacura) [reply projects/hslayers/src/assets/locales/en.json]: We all make mistakes 😉

**Review by fzadrazil — CHANGES_REQUESTED** [review]
Display screenName if first and last name is not filled in.

**Comment by fzadrazil** [comment]
> Display screenName if first and last name is not filled in.

That one works nicely now. But filtering does not. When searching someone by screen name, nothing is found.

File: projects/hslayers/src/common/layman/access-rights/layman-access-rights.html [hunk 3]
```diff
@@ -46,7 +46,7 @@
     <div class="hs-save-map-users-list p-1 my-1">
         <div class="d-flex flex-row justify-content-between align-items-start p-1"
             *ngFor="let user of allUsers | filter: userFilter">
-            <span class="w-50">{{user.name}}</span>
+            <span class="w-50">{{getUserName(user)}}</span>
```
Comment (raitisbe) [inline projects/hslayers/src/common/layman/access-rights/layman-access-rights.html:49]: I don't like function calls inside templates. You never know which one of them becomes slow at some point

File: projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts [hunk 4]
```diff
@@ -1,27 +1,184 @@
...
+  /**
+   * Get all registered users from Layman's endpoint service
+   */
+  async getAllUsers(): Promise<void> {
+    const endpoint: HsEndpoint = this.hsLaymanService.getLaymanEndpoint();
+    if (endpoint?.authenticated) {
+      const url = `${endpoint.url}/rest/users`;
+      try {
+        this.allUsers = await lastValueFrom(
+          this.$http.get<LaymanUser[]>(url, {withCredentials: true}).pipe(
```
Comment (raitisbe) [inline projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts:144]: Very neat

File: projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts [hunk 5]
```diff
@@ -1,27 +1,184 @@
...
+  userFilter = (item): boolean => {
+    const r = new RegExp(this.userSearch, 'i');
+    return r.test(item.name);
+  };
+
+  /**
+   * Refresh user list, when searching for specific user
+   */
+  refreshList(): void {
+    this.allUsers = Array.from(this.allUsers);
```
Comment (raitisbe) [inline projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts:186]: How does this work? Isnt allUsers already an array?

File: projects/hslayers/src/components/save-map/public-api.ts [hunk 6]
```diff
@@ -1,16 +1,16 @@
-export * from './dialog-spawner.service';
-export * from './layer-synchronizer.service';
-export * from './layman-layer-descriptor.interface';
+export * from './feature-services/dialog-spawner.service';
```
Comment (raitisbe) [inline projects/hslayers/src/components/save-map/public-api.ts:1]: I'm not a fan of feature-services directory. Reminds me of having 'partials' directory which was a mistake. Better to group things by function, not form.

File: projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts (hunk 5 above)
Comment (DailisLangovskis) [reply projects/hslayers/src/common/layman/access-rights/layman-access-rights.component.ts:186]: To be honest I don't know, I took it from our code, without it it, doesn't update the list.

File: projects/hslayers/src/common/layman/access-rights/layman-access-rights.html (hunk 3 above)
Comment (DailisLangovskis) [reply projects/hslayers/src/common/layman/access-rights/layman-access-rights.html]: Sorry about this, made in a hurry and forgot to stay out of functions inside loops.

**Review by raitisbe — APPROVED** [review]

**Pull request merged**